### ISSUE TITLE
clean up the salt (reproducibility)

### ISSUE THREAD

**Issue by nesnoj** [openego/data_processing#171]
In DP, we tried to make sure we get the same results for 2 runs using the same data.
This was partially successful:
1. Many crucial data is reproducible, such as demand per load area, ... great! (We did not really compare the data between 2 runs, though. Right?)
2. Some data are not reproducible by design, such as generator positions. Ok!
3. We have **crucial data that are not reproducible yet but should be**

An example of case 3.: The `subst_id`.
Dingo & eDisGo heavily depends on this id as they use it for unique grid numbering. When we create a dataset in dingo using e.g. DP v0.2.10 or use a dataset in eDisGo, we cannot compare to results created with other dp versions without a manual mapping.
This makes life pretty hard at times.

I'd like to fix all of these issues of non-reproducibility but this probably takes a lot of time (this might be a topic for the scenario DB project or eGo2?). For example we could use tables which map the OSM id to the subst_id...

**Quick solution for `subst_id`**
I propose to fix the most critical stuff (for us) in dp (e.g. v0.3.1 or v0.4), which are the ids of MVGD (`subst_id`). The grid district/substation numbering takes place in [`ego_dp_substation_hvmv.sql`](https://github.com/openego/data_processing/blob/c0b360401b6ca0c94e133e54722cbc39216b9bc0/dataprocessing/sql_snippets/ego_dp_substation_hvmv.sql), right? 
What about sorting the OSM ids prior to creation of the substations? This way we'd do it always in the same order, right?
There's one exception: If a new node is added to OSM with an id less than the current max. id, we will get the same problem. But this is 1) quite unlikely since ids are usually ascending and 2) we do not update OSM regularly.

Is this feasible in a short time? @Ludee @ulfmueller @lukasol

**Comment by lukasol** [comment]
I would make use of the osm_id which is always part of the substation tables and does rarely change (as you mentioned). One option would be to simply replace the subst_id by osm_id (probably a bit rough), the other one is to assign an id based on osm_id. We did that for the HV / EHV grid here: 
https://github.com/openego/data_processing/blob/c0b360401b6ca0c94e133e54722cbc39216b9bc0/dataprocessing/sql_snippets/ego_dp_substation_otg.sql#L36-L40

I know it's not really clean since you disregard the osm type, but the risk of having double ids is quite low...
Would something like that solve your problem?

**Comment by gplssm** [comment]
Thanks for your suggestion! In order to ensure uniqueness of the id based on the osm_id we could include the OSM type in building that id (if I got you right). Just to get it right: 
1. Is the osm_id unique?
2. Does the osm_id change for the same substation in a new OSM version?

But, there is another issue: A lot of our DP scripts rely on a range of 1..3608 for the subst_id. All the occurrences would have to be replaced by the new id. I cannot not say how often this occurs and how the refactoring effort would be. Any clues here?

**Comment by lukasol** [comment]
> Is the osm_id unique?

Yes. Well, it is unique for each data type (node, way, relation). The code snippet above shows that we skip the type specification though and expect that the same number does not occur for different types within our substation data set.

> Does the osm_id change for the same substation in a new OSM version?

No. It's unique ;-)

> A lot of our DP scripts rely on a range of 1..3608 for the subst_id. All the occurrences would have to be replaced by the new id. I cannot not say how often this occurs and how the refactoring effort would be. Any clues here?

This would probably mean a restructuring of data types mainly. Depending on the OSM type (node, way relation), the osm_id has different digit lengths (nodes are usually longer than relations). But I think one could do that with reasonable effort.

**Comment by nesnoj** [comment]
Thanks for your contributions!
But what do you think about sorting the OSM ids prior to creation of the substations? (in your initial substation script)
This might can be done with less effort?!

**Comment by lukasol** [comment]
It sounds nice, but I don't really get an idea how you would do that? According to which parameter would you sort the ids?

**Comment by nesnoj** [comment]
Can't we use the way / point ids from osm here?

**Comment by lukasol** [comment]
Yes, you could do that.

**Comment by nesnoj** [comment]
Great. Is `ego_dp_substation_hvmv.sql` the place (and the only place) for these adjustments?

**Comment by lukasol** [comment]
Yes, this is the place where the subst_id is introduced. After that, everything should refer to that.

**Comment by Ludee** [comment]
I prefer the idea to have a `ORDER BY osm_id` instead of replacing the `subst_id`.
Did anybody committed a solution yet?

**Comment by nesnoj** [comment]
Nope...if it's easy for you I'd appreciate if you'd solve this :)

**Comment by Ludee** [comment]
Found some salt in another corner. The zensus for the LA will be ordered now.
Will have a closer look at the substation as well.
